crypto/src/asset/denom.rs

```diff
@@ -0,0 +1,187 @@
+    units: Vec<Unit>,
+}
+
+pub(super) struct Unit {
+    pub exponent: u8,
+    pub denom: String,
+}
+
+impl Inner {
+    /// Constructs the backing data for a set of units.
+    ///
+    /// The base denom is added as a unit, so `units` can be empty and should
+    /// not include a unit for the base denomination.
+    pub fn new(base_denom: String, mut units: Vec<Unit>) -> Self {
+        let id = asset::Id(Fq::from_le_bytes_mod_order(
+            // XXX choice of hash function?
+            blake2b_simd::Params::default()
+                .personal(b"Penumbra_AssetID")
+                .hash(base_denom.as_bytes())
+                .as_bytes(),
+        ));
+
+        for unit in &units {
+            assert_ne!(unit.exponent, 0);
+            assert_ne!(&unit.denom, &base_denom);
+        }
+
+        units.push(Unit {
+            exponent: 0,
+            denom: base_denom.clone(),
+        });
+
+        Self {
+            id,
+            units,
+            base_denom,
+        }
+    }
+}
+
+impl BaseDenom {
+    pub fn id(&self) -> asset::Id {
+        self.inner.id.clone()
+    }
+
+    /// Return a list of display units for this denomination, in priority order.
+    ///
+    /// There will always be at least one display denomination.
+    pub fn units(&self) -> Vec<DisplayDenom> {
+        (0..self.inner.units.len())
+            .map(|unit_index| DisplayDenom {
+                unit_index,
+                inner: self.inner.clone(),
+            })
+            .collect()
+    }
+
+    pub fn default_unit(&self) -> DisplayDenom {
+        self.units()
+            .get(0)
+            .expect("there must be at least one unit");
+
+        DisplayDenom {
+            unit_index: 0,
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+impl From<BaseDenom> for asset::Id {
+    fn from(base: BaseDenom) -> asset::Id {
+        base.id()
+    }
+}
+
+impl Hash for BaseDenom {
+    fn hash<H: hash::Hasher>(&self, state: &mut H) {
+        self.inner.base_denom.hash(state);
+    }
+}
+
+impl PartialEq for BaseDenom {
+    fn eq(&self, other: &Self) -> bool {
+        self.inner.base_denom.eq(&other.inner.base_denom)
+    }
+}
+
+impl Eq for BaseDenom {}
+
+impl Debug for BaseDenom {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.inner.base_denom.as_str())
+    }
+}
+
+impl Display for BaseDenom {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.inner.base_denom.as_str())
+    }
+}
+
+impl DisplayDenom {
+    pub fn base(&self) -> BaseDenom {
+        BaseDenom {
+            inner: self.inner.clone(),
+        }
+    }
+
+    // TODO: API for actually working with displayed denominations?
+    // - format-style printing of amounts of base denom in displaydenom units
+    // - conversion of numbers? (tricky? do you use floats? or what fractions?)
+    //
+    // if we have a good enough api, do we need to have these methods?
```

Comment:
Conversions that are safe:

- `u64` in `DisplayDenom` to `u64` in `BaseDenom` (could error on failure of checked arithmetic if the original quantity is too big to represent, but this is probably `.expect`able)
- `u64` in `BaseDenom` to `String` in `DisplayDenom` (decimal formatted output)
- `String` in `DisplayDenom` (decimal representation) to `u64` in `BaseDenom`

I think we could skip the first one and just do the last two, since this steers arithmetic into being done with the base denomination at all times.